Extend job search for hdcas

mvdbeek opened the pull request:
This PR extends the job search:
  - Jobs that used dataset collections can now be found correctly (jobs have to be newer than this PR, see 19818cb380889e584741caf4a78462e0502f30c1)
  - We check that a input dataset is supplied at the correct input parameter (previously jobs with same inputs for different parameters would be returned as well)
  - The core of the functionality is now in galaxy.manager.jobs so that it can be re-used for https://github.com/galaxyproject/galaxy/issues/4468
  - Some new API tests that exercise this functionality.

mvdbeek committed:
Extend job search for HDCA input

mvdbeek committed:
Check JobParameter match

Matching the JobParameters will prevent returning jobs where a job uses the
same input datasets, but where the inputs are used in a different order (and
hence the result may differ)

mvdbeek committed:
Add job_search tests for other collection types

mvdbeek committed:
Reduce job_search test copypaste

mvdbeek committed:
Allow job_search to find job even if current HDCA was copied from another HDCA and make sure that jobs with same inputs matched to different job parameter inputs are not considered as a match

mvdbeek committed:
Test deleting input datasets and recovering jobs from search

This revealed a wrong query filter (comparing
JobToInputDatasetAssociation.dataset_id == HistoryDatasetAssociation.id,
instead of JobToInputDatasetAssociation.dataset_id ==
HistoryDatasetAssociation.dataset_id).

mvdbeek committed:
Test deleting HDCAs and finding equivalent jobs

mvdbeek committed:
Don't return jobs with deleted HDAs/HDCAs from job_search

mvdbeek committed:
Move job_search out of API for re-use

jmchilton commented on lib/galaxy/jobs/search.py:
```diff
@@ -0,0 +1,143 @@
+from galaxy.managers.lddas import LDDAManager
+
+
+class JobSearch(object):
```
I'd not put this file in ``galaxy.jobs``, I think you are implementing something that belongs in ``galaxy.managers`` somewhere - probably ``galaxy.managers.jobs``. ``galaxy.jobs`` to me feels more like a low level thing related to dealing with actually running jobs. This is implementing manager-style business logic.

mvdbeek committed:
When querying jobs for data input parameters drop all other parameters

mvdbeek committed:
Fix loading of model.JobToInputDatasetCollectionAssociation

If we don't load this relationship we do not properly record a job's
input_dataset_collections in `DefaultToolAction._record_inputs`.

mvdbeek committed:
Filter jobs on input data using JobToInput*Association.

This has the advantage that for complex parameters we don't need to decompose
parameter names like `queries_0|input2` in order to compare JobParameters.

mvdbeek committed:
Move JobSearch to galaxy.managers.jobs

mvdbeek committed:
Return empty list if  type is unknown

mvdbeek committed:
Skip recording JobToInputDatasetCollection relations

if the item to record is not a HistoryDatasetCollectionAssociation. This fixes

```
Error executing tool: Attempting to flush an item of type <class
\'galaxy.model.DatasetCollectionElement\'> as a member of collection
"JobToInputDatasetCollectionAssociation.dataset_collection". Expected an object
of type <class \'galaxy.model.HistoryDatasetCollectionAssociation\'> or a
polymorphic subclass of this type.' ```

It isn't ideal, but before 19818cb we didn't record Collection-like elements at all.

nsoranzo commented on lib/galaxy/managers/jobs.py:
```diff
@@ -0,0 +1,161 @@
+import json
```
Missing blank line here.

nsoranzo commented on lib/galaxy/managers/jobs.py:
```diff
@@ -0,0 +1,161 @@
+
+from galaxy import model
+from galaxy.managers.hdas import HDAManager
+from galaxy.managers.collections import DatasetCollectionManager
```
Wrong import order (you may want to add this file to `.ci/flake8_lint_include_list.txt` so it will be checked)

nsoranzo commented on lib/galaxy/managers/jobs.py:
```diff
@@ -0,0 +1,161 @@
+        return self.__search(tool_id=tool_id, user=user, input_param=input_param, input_data=input_data, job_state=job_state)
+
+    def one_not_deleted(self, items):
+        return any((True for item in items if not item.deleted))
```
You can remove the internal parentheses.

mvdbeek committed:
Linting fixes

and drop unnecessary internal parenthesis (thanks @nsoranzo!).

mvdbeek commented:
Putting this back in WIP, there is probably a problem parsing the input parameters.

mvdbeek commented:
(Mostly writing this down for myself ...)
I'm having some problems with repeats, which can contain a mix of one or more data input parameters and non-data parameters. Repeat parameters are dumped like this and can be retrieved as JobParameter values from the database:
```
'[{"__index__": 0, "f1": {"values": [{"src": "hda", "id": 1}]}}]'
```
This example doesn't contain additional nested repeats, or non-input data, but the problem is that if I query for exactly this JobParameter, I will not be able to find jobs that ran on equivalent input data (A key motivation in https://github.com/galaxyproject/galaxy/issues/4468). Ideally I would like to filter jobs that have this JobParameter, **or** an equivalent JobParameter where the hda id points to the same dataset as the hda with id 1.

My first thought was that instead of querying just for this parameter I could replace the hda id with all equivalent hda ids and instead of using `JobParameter.value == v` I could do `JobParameter.value.in_(v)`. The problem is that this increases in complexity with each hda and data input parameter.

My next idea was to replace the `id` with a wildcard and do `JobParameter.like(v)`, but the `%` wildcard would potentially result in false positives and isn't ideal either.

I am thinking now that I could possibly query for jobs with matching input and output dataset(-collection)s and JobParameters that don't contain a InputDataset. I could then go through the list of possibly equivalent jobs, fill in equivalent dataset ids in the JobParameter dumps and keep only those jobs that match.

mvdbeek committed:
Tighten job search result

We now expand incoming tool parameters, so that we have a dictionary of all
tool parameters (i.e specified/posted parameters and defaults). We start the
search by quering the database for jobs using the same tool id for the same
owner with the same input datasets, taking into account copied
dataset(-collection)s.  We then replace all dataset ids (HDA/LDA or HDCA) in
the parameter dict with the input data used for the jobs queried in the first
step. This allows us to check every parameter for a match with the job.

mvdbeek commented:
Alright, I've implemented my last idea, and it looks like this works pretty well.

jmchilton commented on lib/galaxy/util/__init__.py:
```diff
@@ -68,6 +68,9 @@
 
 
+defaultdict = collections.defaultdict
+
```
I get the impulse here - it is a utility so add it to galaxy.util - but I'd generally prefer dependencies on ``collections`` over dependencies on ``galaxy.util`` in code that consume this. ``galaxy.util`` is very light and doesn't have many nested dependencies but it will never be a portable as ``collections``. I'm not -1 or anything - I'm going to merge this - just sharing my two cents.

mvdbeek replied:
@jmchilton I would have imported from collections, but I'm using `defaultdict` in managers.jobs, but doing `from collections import defaultdict` doesn't work there because we also have managers.collections, this is just a hack around that.

mvdbeek commented:
Thanks for the merge, just saw this now :).
